cloud/scope/powervs_cluster_test.go

```diff
@@ -3043,3 +3043,2423 @@ func TestCreateVPCSubnet(t *testing.T) {
+	})
+
+	t.Run("When SecurityGroupRule doesn't match and return error", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+		rules := infrav1beta2.VPCSecurityGroupRule{
+			Direction: infrav1beta2.VPCSecurityGroupRuleDirectionOutbound,
+			Destination: &infrav1beta2.VPCSecurityGroupRulePrototype{
+				Remotes: append([]infrav1beta2.VPCSecurityGroupRuleRemote{}, infrav1beta2.VPCSecurityGroupRuleRemote{
+					RemoteType: infrav1beta2.VPCSecurityGroupRuleRemoteTypeAny,
+				}),
+				Protocol: infrav1beta2.VPCSecurityGroupRuleProtocolTCP,
+				PortRange: &infrav1beta2.VPCSecurityGroupPortRange{MaximumPort: 65535,
+					MinimumPort: 1,
+				},
+			},
+		}
+		vpcSecurityGroupRule := vpcv1.SecurityGroupRuleSecurityGroupRuleProtocolAll{
+			Direction: ptr.To("outbound"),
+			Protocol:  ptr.To("tcp"),
+			Remote: &vpcv1.SecurityGroupRuleRemote{
+				CIDRBlock: ptr.To("192.168.1.1/24"),
+			},
+			ID: ptr.To("ruleID"),
+		}
+
+		vpcSecurityGroupRules := append([]vpcv1.SecurityGroupRuleIntf{}, &vpcSecurityGroupRule)
+		vpcSecurityGroupStatus := make(map[string]infrav1beta2.VPCSecurityGroupStatus)
+		vpcSecurityGroupStatus["securityGroupName"] = infrav1beta2.VPCSecurityGroupStatus{ID: ptr.To("securityGroupID"), RuleIDs: []*string{}, ControllerCreated: ptr.To(false)}
+		vpcSecurityGroup := infrav1beta2.VPCSecurityGroup{
+			Name:  ptr.To("securityGroupName"),
+			Rules: append([]*infrav1beta2.VPCSecurityGroupRule{}, &rules),
+		}
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient: mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+				Status: infrav1beta2.IBMPowerVSClusterStatus{
+					VPCSecurityGroups: vpcSecurityGroupStatus,
+					VPC: &infrav1beta2.ResourceReference{
+						ID: ptr.To("VPCID"),
+					},
+				},
+				Spec: infrav1beta2.IBMPowerVSClusterSpec{
+					VPCSecurityGroups: append([]infrav1beta2.VPCSecurityGroup{}, vpcSecurityGroup),
+				},
+			},
+		}
+
+		securityGroupDetails := &vpcv1.SecurityGroup{Name: ptr.To("securityGroupName"), ID: ptr.To("securityGroupID"), Rules: vpcSecurityGroupRules, VPC: &vpcv1.VPCReference{ID: ptr.To("VPCID")}}
+		mockVPC.EXPECT().GetSecurityGroupByName(gomock.Any()).Return(securityGroupDetails, nil)
+		sg, ruleIDs, err := clusterScope.validateVPCSecurityGroup(vpcSecurityGroup)
+		g.Expect(ruleIDs).To(BeNil())
+		g.Expect(sg).To(BeNil())
+		g.Expect(err).ToNot(BeNil())
+	})
+}
+
+func TestValidateVPCSecurityGroupRule(t *testing.T) {
+	var (
+		mockVPC  *mock.MockVpc
+		mockCtrl *gomock.Controller
+	)
+	setup := func(t *testing.T) {
+		t.Helper()
+		mockCtrl = gomock.NewController(t)
+		mockVPC = mock.NewMockVpc(mockCtrl)
+	}
+	teardown := func() {
+		mockCtrl.Finish()
+	}
+
+	t.Run("When it matches SecurityGroupRule of protocolType SecurityGroupRuleSecurityGroupRuleProtocolTcpudp", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+		remote := infrav1beta2.VPCSecurityGroupRuleRemote{
+			CIDRSubnetName: ptr.To("CIDRSubnetName"),
```

Comment:
You should not use `CIDRSubnetName` when its of RemoteTypeAny. It should be used only `RemoteTypeCIDR`